Bigquery snippets

davidcavazos opened the pull request:
@tvalentyn Java
@aaltay Python

Added Java BigQueryIO snippets and updated the Python version to match.

davidcavazos commented:
@melap

tvalentyn commented on examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java:
```diff
@@ -0,0 +1,233 @@
+        tableSpec = writeProject + ":" + writeDataset + "." + writeTable;
+      }
+
+      // [START BigQueryWrite]
```
<!--new_thread; commit:9eba19b98f4e946b4d764bd60be0b2a39d3a8124; resolved:0-->
Do we need the commented part?

davidcavazos replied:
It's not needed, but as part of a snippet, it's good for the reader to know the whole context :)

tvalentyn commented on sdks/python/apache_beam/examples/snippets/snippets.py:
```diff
@@ -989,46 +989,93 @@ def to_entity(content):
+      {'source': 'Mahatma Ghandi', 'quote': 'My life is my message.'}
+  ])
+
+  quotes | beam.io.Write(beam.io.BigQuerySink(
```
<!--new_thread; commit:9eba19b98f4e946b4d764bd60be0b2a39d3a8124; resolved:0-->
AFAIK beam.io.WriteToBigQuery is a better choice than BigQuerySink and we should change examples to use that. Filed https://issues.apache.org/jira/browse/BEAM-3435 as a placeholder.

davidcavazos replied:
Got it, will update. Thanks!

davidcavazos commented:
Ready for review

jkff commented on examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java:
```diff
@@ -0,0 +1,233 @@
+
+  @DefaultCoder(AvroCoder.class)
+  static class Quote {
+    String source;
```
Can these be final?

jkff commented on examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java:
```diff
@@ -0,0 +1,233 @@
+          .apply(BigQueryIO.readTableRows().from(tableSpec))
+          // Each row is of type TableRow
+          .apply(MapElements.into(TypeDescriptors.doubles()).via(
+              (TableRow row) -> (Double) row.get("max_temperature")));
```
Nice, thanks for using Java8 here.

jkff commented on examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java:
```diff
@@ -0,0 +1,233 @@
+
+    {
+      // [START BigQuerySchema]
+      String tableSchema = "source:STRING, quote:STRING";
```
I suppose this refers to usage of .withJsonSchema(String jsonSchema) - but this string is not in JSON format. What's up with this?

davidcavazos replied:
Good catch, I copied-pasted the string format from Python. Apparently they are not compatible between Java and Python. Modifying to use actual JSON. Thanks

jkff commented on examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java:
```diff
@@ -0,0 +1,233 @@
+
+    {
+      String tableSpec = "clouddataflow-readonly:samples.weather_stations";
+      if (writeProject != "" && writeDataset != "" && writeTable != "") {
```
What if one of them is empty but others aren't?

jkff commented on examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java:
```diff
@@ -0,0 +1,233 @@
+      quotes
+          .apply(MapElements.into(TypeDescriptor.of(TableRow.class)).via(
+              (Quote elem) -> {
+                TableRow row = new TableRow();
```
This can be abbreviated to new TableRow().set(...).set(...)

jkff commented on examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java:
```diff
@@ -0,0 +1,233 @@
+  }
+
+  /** Helper function to format results in coGroupByKeyTuple. */
+  public static String formatCoGbkResults(
```
Is this related to bigquery snippets?

davidcavazos replied:
No, I renamed the directory and git apparently took it as replacing it with a new file. This part is unchanged.

jkff commented on examples/java8/src/test/java/org/apache/beam/examples/snippets/SnippetsTest.java:
```diff
@@ -42,7 +45,29 @@
+      p.run();
+    } else {
+      Pipeline p = Pipeline.create();
+      Snippets.modelBigQueryIO(p);
```
p.run() ?

davidcavazos replied:
Running this will fail due to credentials and table permissions. Therefore we limit to making sure the transforms can be built and that every validation passes. See comment at line 52.

jkff commented on examples/java8/src/test/java/org/apache/beam/examples/snippets/SnippetsTest.java:
```diff
@@ -42,7 +45,29 @@
+  public void testModelBigQueryIO() {
+    // We cannot test BigQueryIO functionality in unit tests, therefore we limit ourselves
+    // to making sure the pipeline containing BigQuery sources and sinks can be built.
+    boolean runLocally = false;
```
Are lines 56-65 dead code?

davidcavazos replied:
Basically yes, but they are useful when testing locally. Replacing the `project`, `dataset` and `table` variables with whatever you're using allows for testing locally and making sure everything is doing what it's supposed to do

davidcavazos commented:
Updated Java snippets to address comments, ready for review

chamikaramj commented on examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java:
```diff
@@ -0,0 +1,248 @@
+      */
+
+      PCollection<Quote> quotes = p.apply(Create.of(
+          new Quote("Mahatma Ghandi", "My life is my message.")
```
<!--new_thread; commit:e6e450ec518edfbcd98c912cdeae9998504b78ce; resolved:0-->
s/Ghandi/Gandhi :)

davidcavazos replied:
Thanks! Changing

chamikaramj commented on sdks/python/apache_beam/examples/snippets/snippets.py:
```diff
@@ -989,46 +989,93 @@ def to_entity(content):
+  table_schema.fields.append(quote_field)
+  # [END model_bigqueryio_schema_object]
+
+  if write_project and write_project and write_table:
```
<!--new_thread; commit:bc82062e43687f58c37b92da05e37dd607635fd0; resolved:0-->
write_project and write_dataset ...

davidcavazos replied:
Good catch, thanks!

chamikaramj commented on sdks/python/apache_beam/examples/snippets/snippets_test.py:
```diff
@@ -589,18 +590,30 @@ def test_model_textio_compressed(self):
     # ourselves to making sure the pipeline containing BigQuery sources and
     # sinks can be built.
-    snippets.model_bigqueryio()
+    run_locally = False
```
<!--new_thread; commit:9d7586c0b982dd7eab8f3e3bbc28864792629dda; resolved:0-->
Why do we have to branch below given that this doesn't change ?

davidcavazos replied:
To make it easier to test locally, I added a comment clarifying this.

chamikaramj commented on examples/java8/src/test/java/org/apache/beam/examples/snippets/SnippetsTest.java:
```diff
@@ -42,7 +45,29 @@
+  public void testModelBigQueryIO() {
+    // We cannot test BigQueryIO functionality in unit tests, therefore we limit ourselves
+    // to making sure the pipeline containing BigQuery sources and sinks can be built.
+    boolean runLocally = false;
```
<!--new_thread; commit:ae660d95c3a81daa41104384866f652ae761a937; resolved:0-->
Why do we need branching below given that this is always false ?

davidcavazos replied:
It just makes it easier for us to test locally. I'll add some instructions as comments to make it clear.

davidcavazos commented:
Addressed review comments, ready for review. As for the "run locally" logic, let me know if you would prefer to just remove that. I left it there for ease of locally running the pipeline and making sure everything actually ran correctly since what we currently run is only to make sure the transforms can be built and all validations pass, but the pipeline itself doesn't run.

jkff commented on examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java:
```diff
@@ -0,0 +1,248 @@
+      String tableSpec = "clouddataflow-readonly:samples.weather_stations";
+      // [START BigQueryReadTable]
+      PCollection<Double> maxTemperatures = p
+          .apply(BigQueryIO.readTableRows().from(tableSpec))
```
Suggestion: `readTableRows()` has poor performance - consider including also a snippet of `read(SerializableFunction)` which performs much better. It's also definitely a better fit for what you're doing here (immediately extracting some information from the row). In fact I'd recommend it for *all* snippets reading from BigQuery, and maybe include just 1 snippet with readTableRows.

davidcavazos replied:
Got it, changing

davidcavazos commented:
Updated to use `read(SerializableFunction)`. Ready for review :)

jkff commented:
Java snippets LGTM; deferring to Cham on the Python snippets.

Reiterating previous question: when shall we get snippets for BigQuery writes with DynamicDestinations? In a separate PR or should that be done here?

davidcavazos commented:
I can include them on this PR, I'll work on them these next days :)

davidcavazos commented:
Dynamic Destinations snippet added, ready for review

jkff commented on examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java:
```diff
@@ -0,0 +1,385 @@
+                  new TableReference()
+                      .setProjectId(writeProject)
+                      .setDatasetId(writeDataset)
+                      .setTableId(writeTable + "_" + destination),
```
In terms of BigQuery best practices, partitioning by time is better done using partition decorators (`table$YYYYMMDD`).

davidcavazos commented:
Rebased to latest version, changed `SerializableCoder.of(Double.class)` to `DoubleCoder.of()`. Python tests are broken, see [BEAM-3628](https://issues.apache.org/jira/browse/BEAM-3628). Ready for review.

davidcavazos commented:
@chamikaramj does this look good to you? I think it's basically ready to merge unless there are any other comments, thanks!

chamikaramj commented:
LGTM for Python. Thanks.

Merging.

chamikaramj commented:
Please do a git fix-up and merge commits to single/few commit(s).

davidcavazos commented:
Commits merged

chamikaramj commented:
Run Java PreCommit
